hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -501,7 +501,17 @@ private Boolean destroyCurrentClusterConnectionAndTryNextCluster(CandidateCluste
         logger.info("Trying to connect to next cluster: " + nextContext.getClusterName());
 
         if (doConnectToCandidateCluster(nextContext, true)) {
-            client.waitForInitialMembershipEvents();
```

Comment:
I think that a better solution is to wait for the latch + if the connection is disconnected, stop waiting. When member dies we should get it detected via connection lost and take necessary cleanup and fail the connection attempt.

Reply:
but it can happen in other cases than disconnection. For example a network delay or a very long gc in the member side can cause the same thing.

Reply:
That's why I check if the wait on initial membership event fails

Reply:
If connection is closed, no need to wait for the timeout, but do we complete the latch if the connection is closed without waiting for the timeout???

Reply:
no we don't. if we start waiting, it'll block

Reply:
I did not mean to wait, I meant to complete the latch so that it will not wait on the latch.

Reply:
What you suggest is not related to what I'm fixing but another improvement we can implement as I understand. Not sure how I can implement, we need to interrupt the waiting thread and somehow say it that it should not wait anymore. 

any idea how to implement that? I'd like to do that in another PR since it's not related to my fix and I am not clear about how to do that.